Add Cloud Optimized GeoTIFF Support

joshfix opened the pull request:
This PR introduces a new plugin that seeks to take advantage of the COG structure by asynchronously reading byte ranges for contiguous TIFF tiles.  It includes caching and non-caching ImageInputStream implementations, as well as clients (RangeReaders) capable of reading remote COGs using HTTP, Azure, and AWS S3 client libraries.

dromagnoli commented on plugin/cog/cog-rangereader-http/src/main/java/it/geosolutions/imageioimpl/plugins/cog/HttpRangeReader.java:
```diff
@@ -0,0 +1,195 @@
+        try {
+            buffer.put(bytes);
+        } catch (Exception e) {
+            LOGGER.severe("Error writing bytes to ByteBuffer for source " + uri);
```
Wondering if wee need to also log the causing exception in the error message.

dromagnoli commented on plugin/cog/cog-rangereader-s3/src/main/java/it/geosolutions/imageioimpl/plugins/cog/S3RangeReader.java:
```diff
@@ -0,0 +1,174 @@
+                            writeValue((int)key, future.get().asByteArray());
+                            completed.add(key);
+                        } catch (Exception e) {
+                            e.printStackTrace();
```
See my comments above on the "XYZRangeReader" about printing StackTrace

dromagnoli commented on plugin/cog/cog-streams/src/main/java/it/geosolutions/imageioimpl/plugins/cog/DefaultCogImageInputStream.java:
```diff
@@ -0,0 +1,399 @@
+        try {
+            delegate.seek(streamPos);
+        } catch (IOException e) {
+            e.printStackTrace();
```
Here too :)

dromagnoli commented on plugin/cog/cog-streams/src/main/java/it/geosolutions/imageioimpl/plugins/cog/CachingCogImageInputStream.java:
```diff
@@ -0,0 +1,244 @@
+                        Arrays.copyOfRange(b, (int) tileRange.getStart(), (int) (tileRange.getEnd() + 1));
+                CacheManagement.DEFAULT.cacheTile(key, tileBytes);
+            } catch (Exception e) {
+                e.printStackTrace();
```
Same as above on printing stack trace

joshfix commented:
I believe I have addressed all of the issues pointed out.  Please let me know if there are more findings and I will address those as well.

joshfix commented on plugin/cog/cog-rangereader-azure/src/main/java/it/geosolutions/imageioimpl/plugins/cog/AzureRangeReader.java:
```diff
@@ -0,0 +1,190 @@
+
+                            completed.add(key);
+                        } catch (Exception e) {
+                            e.printStackTrace();
```
If this fails, 1 or more tiles may become corrupt.  This could lead to an error decoding that tile(s).  Because TIFFImageReader reads 1 tile at a time, I wasn't 100% certain if it would fail on a single failed tile, or if only that tile would not render.  Being optimistic, I didn't fail hard -- I wanted to give the other ranges/tiles a chance to be read and cached in the event they still might be rendered.

dromagnoli commented:
Thanks @joshfix 
@aaime will do a review too early next week.

aaime commented on plugin/cog/cog-streams/src/main/java/it/geosolutions/imageioimpl/plugins/cog/CogImageInputStreamSpi.java:
```diff
@@ -0,0 +1,82 @@
+            }
+        }
+        if (input instanceof String || input instanceof URL || input instanceof URI) {
+            if (((CogUri)input).isUseCache()) {
```
This seems to guarantee an ClassCastException, given the input might only be a string, url or uri at this point. Maybe in this if there should be some sort of default configuration based on system properties, that would make it choose the caching or not caching path?

joshfix replied:
Nice catch... the other solution would be to simply not support string/url/uri.  I can already see this being an issue when trying to lookup stream implementations as other input streams already advertise string/url etc as their input classes.  Maybe I should keep this class as only accepting `CogUri` so there is no confusion?

aaime replied:
I'm afraid it's not going to work in GeoServer if strings/URLs are not accepted, but I may be wrong (or it might be that we can build a CogUri there, just don't think it's possible to do so now, if memory serves me right every reader works off java own URL).

aaime commented on plugin/cog/cog-streams/src/main/java/it/geosolutions/imageioimpl/plugins/cog/DefaultCogImageInputStream.java:
```diff
@@ -0,0 +1,399 @@
+ * @author joshfix
+ * Created on 2019-08-23
+ */
+public class DefaultCogImageInputStream implements ImageInputStream, CogImageInputStream {
```
So this class works in two steps, initially it has no content and won't work, then one inits the header, and it can provide the header only, and after calling readranges, it can offer the tiles too (only the requested ones). So it's not a generic image input stream, but a special purpose one that can only be used in concert with the COG reader. This should be explicit imho.

joshfix replied:
Also all correct.  The central concept of my design is to "warm" the streams with the data of (only) the requested tiles up front, before we actually start to decode tiles.  The only way that can be accomplished is by reading the header to determine the tile offsets and lengths.  Basically the input streams are completely non-functional without being initialized with the header data. 

I guess I assumed from the name and the interface it would be apparent that it was purpose-built for COG, but I will add more documentation stating and explaining its purpose and limitations.

aaime replied:
ImageInputStream are an extension point that can be looked up and used in whatever other format, making one that's specific is off its intended usage. I understand the reason behind the current implementation, and hopefully it's not going to cause side effects for other concurrent sources in the same server (e.g., other geotiffs, ecws and the like), but it's best to be quite clear when going off design.

joshfix replied:
Added explicit documentation regarding purpose.

aaime commented on plugin/cog/cog-commons/src/main/java/it/geosolutions/imageioimpl/plugins/cog/RangeReader.java:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    public void setFilesize(int filesize) {
+        buffer = ByteBuffer.allocate(filesize);
```
Wow, danger zone! The byte buffer here is actually allocated in memory, so if one has to deal with a bigtiff that's say 100GB in the cloud, the code will actually allocate in the local memory 100GB, even if the current read only aims at reading one tile?
See: https://docs.oracle.com/javase/7/docs/api/java/nio/ByteBuffer.html#allocate(int)

If I understand correctly, the current implementation works for (very) small images and light load, but it's deadly easy to make it go OOM. It would need some sort of sparse byte buffer, or, instead of delegating to a MemoryCacheImageInputStream, build its own implementation that only keeps in memory the tiles needed for the current read... should be doable by keeping an "index" of offsets pointing at the actual bytes, and doing some offset math whenever a read is called.

joshfix replied:
Thanks for pointing this out.  In fact I did not take into account very large files or consider the ramifications of blindly allocating the full filesize in the byte buffer.  I'm going to spend a bit of time contemplating the best way to handle this.  My initial design was to keep the `RangeReader` implementations responsibilities as narrow as possible -- read the final ranges required for the requested tiles, but know nothing about tiles, then fill in the byte buffer at the appropriate places.  I'm thinking maybe this should be modified so that:

* The `CogTileInfo` object is passed to the range reader so that it has full knowledge of each requested tile index, offset, length, etc
* The range reader uses the `RangeBuilder` to build the contiguous ranges to be read
* The range reader then decomposes the results back into per-tile byte arrays (possibly storing them in the corresponding `TileRange` objects in `CogTileInfo`
* Update the `DefaultCogImageInputStream` accordingly with offset math, etc

What do you think about this solution?

aaime replied:
I think the RangeReader classes can be kept compact and focused, if instead of returning a byte[], end up returning a ``Map<long[], byte[]>`` with just the byte ranges read. The interesting and bit complicated bit will be updating the ``CogImageInputStream`` implementations so that they keep track of the current reading position and do the offset math dance. I don't see ``TileInfo`` needing to be updated, but it's more of an implementation preference.

joshfix replied:
Updated to return `Map<Long, byte[]>`, where the Long value is the start of the range.  The end of the range, if needed, can be calculated using the byte[] length.

joshfix replied:
Oh, I suppose it's important to mention that I completely removed a) any reference or attempt to understand the actual file size and b) usage of the class-level ByteBuffer to store all range data :)

aaime commented on plugin/cog/cog-rangereader-http/src/main/java/it/geosolutions/imageioimpl/plugins/cog/HttpRangeReader.java:
```diff
@@ -0,0 +1,194 @@
+                String length = contentRange.split("/")[1];
+                try {
+                    filesize = Integer.parseInt(length);
+                    buffer = ByteBuffer.allocate(filesize);
```
OOM issue, another problematic point.

joshfix replied:
ByteBuffer has been removed.

aaime commented on plugin/cog/cog-rangereader-s3/src/main/java/it/geosolutions/imageioimpl/plugins/cog/S3RangeReader.java:
```diff
@@ -0,0 +1,175 @@
+        try {
+            HeadObjectResponse headResponse = client.headObject(headObjectRequest).get();
+            filesize = headResponse.contentLength().intValue();
+            buffer = ByteBuffer.allocate(filesize);
```
OOM issue, another problematic point.

joshfix replied:
ByteBuffer has been removed.

aaime commented on plugin/cog/cog-rangereader-azure/src/main/java/it/geosolutions/imageioimpl/plugins/cog/AzureRangeReader.java:
```diff
@@ -0,0 +1,182 @@
+        connector = new AzureConnector(AzureUrlParser.getAccountName(uri));
+        client = connector.getAzureClient(containerName, filename);
+        filesize = (int) client.getProperties().block().blobSize();
+        buffer = ByteBuffer.allocate(filesize);
```
OOM issue, another problematic point.

joshfix replied:
ByteBuffer has been removed.

joshfix commented on plugin/cog/pom.xml:
```diff
@@ -0,0 +1,92 @@
+                <version>${azure.version}</version>
+            </dependency>
+            <dependency>
+                <groupId>com.squareup.okhttp3</groupId>
```
OkHttp seems to be a pretty popular client, but I'm not tied to it.  This was actually a rewrite as I originally used the Java 11 HttpClient, but Torben pointed out that this needs to be compatible with Java 8, so I moved to OkHttp.  I have no issue moving to commons-httpclient as it seems to support async connections.  My question now is, would you be against introducing v5 of the client?  It supports http/2 multiplexing, which eliminates the concerns of how many connections get created.  This seems like the best modern approach to take, but still introduces a new dependency.  Or potentially I could provide an http1 and http2 client?

https://hc.apache.org/httpcomponents-client-5.0.x/httpclient5/examples/AsyncClientH2Multiplexing.java

aaime replied:
If it's not aligned with the GeoTools current usage it would be one more dependency nevertheless, at that point, probably better to keep it as is and concentrate on the other issues.

joshfix replied:
Left it alone for now.  I'm fine with rewriting this in the future or refactoring this to be the http2 module and providing another http module based on commons-httpclient.

joshfix commented on plugin/cog/cog-commons/src/main/java/it/geosolutions/imageioimpl/plugins/cog/CogTileInfo.java:
```diff
@@ -0,0 +1,96 @@
+        this.headerSize = headerSize;
+    }
+
+    public void addTileRange(int tileIndex, long offset, long byteLength) {
```
Could you provide a bit more detail about your concern here?  Is it just that the two separate if statements are confusing?  I will definitely take another pass at documenting methods better.

aaime replied:
It's one method call that does two entirely different things based on input parameters, at the very least they should be explained. Or just have two separate methods with clear names instead (javadocs are still a plus in this case too).

joshfix replied:
Removed tracking the start tile index, added documentation about the remaining logic to check header size.

joshfix commented on plugin/cog/cog-commons/src/main/java/it/geosolutions/imageioimpl/plugins/cog/RangeReader.java:
```diff
@@ -0,0 +1,109 @@
+        return headerLength;
+    }
+
+    public abstract void readAsync(long[]... ranges);
```
All correct.  Looks like I'll be refactoring all of this to avoid the usage of ByteBuffer.  Had async requests stuck in my brain when I created that method, and originally I also had another read method to make a single request that was purely synchronous.  I'll just rename the method as it is in fact, non-synchronous and will return the result after I refactor.

joshfix replied:
Updated the name and modified the method to return the data.

joshfix commented on plugin/tiff/src/main/java/it/geosolutions/imageioimpl/plugins/tiff/TIFFImageReaderSpi.java:
```diff
@@ -125,6 +125,10 @@ public TIFFImageReaderSpi() {
               );
     }
 
+    public static void setReaderClassName(String readerClassName) {
```
Updated to make this non-static.

joshfix commented on plugin/cog/cog-commons/src/main/java/it/geosolutions/imageioimpl/plugins/cog/CogImageInputStream.java:
```diff
@@ -0,0 +1,36 @@
+ */
+public interface CogImageInputStream {
+
+    void readRanges();
```
All of the interfaces in the commons module are now thoroughly commented at the method level.

joshfix commented:
Committed changes that should address all concerns @aaime.  In addition, I made a few other noteworthy changes:

* The azure sdk was a preview version and I started encountering very odd behavior.  I have decided to remove support entirely until the library stabilizes.
* `RangeBuilder` was renamed to `ContiguousRangeComposer` to help further distinguish it from `RangeReader` naming.
*  Thread/connection pools (and the ability to configure them via environment variables or system properties) has been added to the S3 and HTTP range reader implementations.
*  File size caching has been removed, as the code no longer attempts to obtain or use the full file size in any capacity.
* The mime type "image/tiff; application=geotiff" has been added to `TIFFImageReaderSpi` (see: https://github.com/opengeospatial/geotiff/issues/34#issuecomment-514078289)

joshfix commented on plugin/cog/cog-reader/src/main/java/it/geosolutions/imageioimpl/plugins/cog/CogImageReader.java:
```diff
@@ -0,0 +1,124 @@
+
+/**
+ * ImageReader implementation extending from TIFFImageReader.  If this class encounters an ImageInputStream that does
+ * not implement `CogImageInputStream`, it will simply pas the request on to TIFFImageReader.  Otherwise, it will
```
fixed

joshfix commented on plugin/cog/cog-commons/src/main/java/it/geosolutions/imageioimpl/plugins/cog/CogTileInfo.java:
```diff
@@ -0,0 +1,147 @@
+    protected int headerLength = DEFAULT_HEADER_LENGTH;
+    protected Map<Integer, TileRange> tileRanges = new TreeMap<>();
+    public static final int HEADER_TILE_INDEX = -100;
+    public static final int DEFAULT_HEADER_LENGTH = 16384;
```
I personally have not seen this and decided on 16K because it's the default for GDAL COG reads as well.  I would assume if a TIFF had enough overviews and enough tiles, you could eventually break past 16K, but it seems unlikely to me?  I just realized that I made an effort to make the initial headerLength configurable, but provided no real mechanism to allow an external user to set it.  I'm thinking I could just add it as a field to `CogImageReadParam`.... thoughts?

joshfix commented on plugin/cog/cog-commons/src/main/java/it/geosolutions/imageioimpl/plugins/cog/RangeReader.java:
```diff
@@ -0,0 +1,54 @@
+package it.geosolutions.imageioimpl.plugins.cog;
```
re-added

joshfix commented on plugin/cog/cog-streams/src/main/java/it/geosolutions/imageioimpl/plugins/cog/DefaultCogImageInputStream.java:
```diff
@@ -0,0 +1,187 @@
+
+        // this should never happen -- we should have read all bytes from all tiles in the request envelope
+        if (contiguousRange == null || rangeStart == -1L) {
+            streamPos += len;
```
updated to throw an exception

joshfix commented on plugin/cog/cog-streams/src/main/java/it/geosolutions/imageioimpl/plugins/cog/DefaultCogImageInputStream.java:
```diff
@@ -0,0 +1,187 @@
+
+        int relativeStreamPos = (int)(streamPos - rangeStart) + off;
+        // copy the bytes from the fetched tile into the destination byte array
+        for (int i = 0; i < len; i++) {
```
updated this and the caching implementation to use System.arraycopy.

joshfix commented:
Re: how to test in GeoServer...  There is a blocker :)

I think we would really just want the ability to use the GeoTools GeoTiffReader.  The problem is that the GeoTiffFormat class doesn't allow you to modify the image reader SPI:

https://github.com/geotools/geotools/blob/master/modules/plugin/geotiff/src/main/java/org/geotools/gce/geotiff/GeoTiffFormat.java#L127

To circumvent the issue, I created my own format class that is nearly an exact duplicate of GeoTiffFormat, but the `getReader` method always returns a COG reader:

https://github.com/joshfix/cog-image-reader/blob/master/src/main/java/com/joshfix/geotools/geotiff/CogFormat.java#L299

(note this is utilizing an old version of the COG project)

I have tested this in the past using a custom mosaic store that uses a STAC catalog as a feature source and uses the aforementioned format class as the suggested format:

https://github.com/joshfix/imagemosaic-stac/blob/dev/src/main/java/com/joshfix/stac/store/mosaic/StacMosaicReader.java#L267

(forgive the ugly code, i just pushed the dev branch which was full of in-progress tests)  It worked fine for me and I was able to dynamically render landsat 8 mosaics.  Not sure what the best/recommended approach would be here, but I think we're going to need to update some GeoTools code to get this working in GeoServer.

aaime commented:
Doh sorry I missed your earlier comment. How about modifying the GeoTIFF format to use a list of potential image readers instead? And skip the subclass?

joshfix commented:
No worries!  I think you may have missed my other comment/question as well about setting the header length.  I realized that even though most of my classes accept header length as a variable, there is no way outside of the CogImageReader to set it.  Previously, the default value came from `CogTileInfo`.  I modified the code to make the value come from CogImageReadParam.  It still has the same default value, but now it's possible to pass in a custom value via the param.  I also pushed a change that overrides `canDecodeInput` in the image reader SPI to make sure the stream is initialized.

RE: modifying GeoTiffFormat... Given that it's a GeoTools class, does that mean that this imageio-ext code is good to go?  Should we move this conversation now to GeoTools?  I am definitely open to suggestion here on how to handle the GeoTools code. 

It seems we would need to deal with this in GeoTiffFormat:

```
    /** SPI for the reader. */
    private static final TIFFImageReaderSpi IMAGEIO_READER_FACTORY = new TIFFImageReaderSpi();
```

GeoTiffReader has the same private static final field. In addition, there may be some situations we need to handle.  If a `CogUrl` is passed as input to GeoTiffReader, ImageIOExt.getImageInputStreamSPI should be able to create a new CogImageInputStream, however it will not be initialized (the RangeReader implementation will not be set and the header will not be read).  This also applies to anywhere the reader is calling `READER_SPI.createReaderInstance()`... we'll need to handle initializing it if it's a Cog image reader.  Other things that could be problematic: the `getHRInfo` uses the SPI to create a reader instance, then calls `reader.getImageMetadata`, which will suffer from the same problem of the stream not being initialized.  A Hints object is passed around in GeoTiffReader, in the constructor, and in getHRInfo, so maybe that would be something we could take advantage of to help handle COG readers?

Sorry for the long-winded comment.  Let me know how you'd like to proceed from here!

aaime commented:
Hi @joshfix so I wanted to go ahead and merge, and then move attention to GT/GS indeed.
Tried to build locally, was greeted by this:

```
Tests in error: 
  readHttpRanges(it.geosolutions.imageio.cog.S3RangeReaderTest): java.util.concurrent.ExecutionException: software.amazon.awssdk.services.s3.model.S3Exception: The AWS Access Key Id you provided does not exist in our records. (Service: S3, Status Code: 403, Request ID: 5EDD5F80B931B58F)
  readS3Ranges(it.geosolutions.imageio.cog.S3RangeReaderTest): java.util.concurrent.ExecutionException: software.amazon.awssdk.services.s3.model.S3Exception: The AWS Access Key Id you provided does not exist in our records. (Service: S3, Status Code: 403, Request ID: 8BD251F71DD2C191)
```

If there are online tests that need specific setups, or even just internet access, it's best if they are enabled only with a profile, or a config file that the dev has to provide.

joshfix commented:
This is very interesting.  I can't reproduce locally and the test runs against a publically available landsat 8 image on s3, so no aws key is required for to read the image.

aaime commented:
Hum... a test that requires an internet connection should still be activated only by profiles.
About the failures, maybe I found something... I have an old .aws directory in my home, that has credentials that are probably expired? I tried removing it, and the error message changes:

```

Tests in error: 
  readHttpRanges(it.geosolutions.imageio.cog.S3RangeReaderTest): java.util.concurrent.ExecutionException: software.amazon.awssdk.core.exception.SdkClientException: Unable to load credentials from any of the providers in the chain AwsCredentialsProviderChain(credentialsProviders=[SystemPropertyCredentialsProvider(), EnvironmentVariableCredentialsProvider(), ProfileCredentialsProvider(), ContainerCredentialsProvider(), InstanceProfileCredentialsProvider()]) : [SystemPropertyCredentialsProvider(): Unable to load credentials from system settings. Access key must be specified either via environment variable (AWS_ACCESS_KEY_ID) or system property (aws.accessKeyId)., EnvironmentVariableCredentialsProvider(): Unable to load credentials from system settings. Access key must be specified either via environment variable (AWS_ACCESS_KEY_ID) or system property (aws.accessKeyId)., ProfileCredentialsProvider(): Profile file contained no credentials for profile 'default': ProfileFile(profiles=[]), ContainerCredentialsProvider(): Cannot fetch credentials from container - neither AWS_CONTAINER_CREDENTIALS_FULL_URI or AWS_CONTAINER_CREDENTIALS_RELATIVE_URI environment variables are set., InstanceProfileCredentialsProvider(): Unable to load credentials from service endpoint.]
  readS3Ranges(it.geosolutions.imageio.cog.S3RangeReaderTest): java.util.concurrent.ExecutionException: software.amazon.awssdk.core.exception.SdkClientException: Unable to load credentials from any of the providers in the chain AwsCredentialsProviderChain(credentialsProviders=[SystemPropertyCredentialsProvider(), EnvironmentVariableCredentialsProvider(), ProfileCredentialsProvider(), ContainerCredentialsProvider(), InstanceProfileCredentialsProvider()]) : [SystemPropertyCredentialsProvider(): Unable to load credentials from system settings. Access key must be specified either via environment variable (AWS_ACCESS_KEY_ID) or system property (aws.accessKeyId)., EnvironmentVariableCredentialsProvider(): Unable to load credentials from system settings. Access key must be specified either via environment variable (AWS_ACCESS_KEY_ID) or system property (aws.accessKeyId)., ProfileCredentialsProvider(): Profile file contained no credentials for profile 'default': ProfileFile(profiles=[]), ContainerCredentialsProvider(): Cannot fetch credentials from container - neither AWS_CONTAINER_CREDENTIALS_FULL_URI or AWS_CONTAINER_CREDENTIALS_RELATIVE_URI environment variables are set., InstanceProfileCredentialsProvider(): Unable to load credentials from service endpoint.]
```

Rings any bell? Could it be that you're not seeing the error as you have the "right" credentials or at least a set of valid ones?

joshfix commented:
I renamed my aws config files and was able to reproduce.  It was a simple fix -- when I'm building the s3 client, if no credentials were provided I build an anonymous provider.  I also added an online profile for all tests that require internet access.

aaime commented:
Merged finally! Going to have a look at the geotools/geoserver side

aaime commented:
Ok had a look, you're right, having to initialize the streams with the range reader is source of troubles indeed... at this point I don't see much alternative to making a custom subclass of the GeoTiff reader.

There you can possibly dodge the SPI mechanism and manually setup the stream, and init it using the desired area to read and target resolution (which are available in the read method of the reader).

If no area is required, then the whole tiff at max resolution is read, to just read the metadata, make it read only the header I believe?

I would support making changes to the base gt-geotiff classes, enough to make it possible to setup the streams correctly.

Does this sound like a plan?

joshfix commented:
Thanks for getting this merged!  Very excited to take the next step and get this working in GeoTools.  I'm a bit buried at the moment so it may be a bit before I can circle back to this.  Totally open to any suggestions, however, and if you see a clear path forward I'll be happy to implement it.
